### DISCUSSION TITLE
Action before deleting an object

### DISCUSSION THREAD

**Discussion by mcarella** [evennia/evennia#2403] [Evennia Questions] [answered]
Hey :-)

My problem is this: I have an object let's say an NPC carrying a bunch of other items.
When I delete it, all the items contained into the NPC goes to its own Home location (limbo).

I want the NPC drop the items before getting destroyed:

This piece of code I've putted into NPC class seems not working:

```
def dropandie(self):

     for item in self.contents:
          item.move_to(self.location, quiet=True)
     
     self.delete()
```

it's like the self.delete() it is called before the first block of code, that's why if I comment the first two rows , it seems working. 

I've also tried using the `at_object_delete()` without no success.

Any tips for helping me? 🙏 
Thank you!

**Comment by kovitikus** [comment]
```py
if len(self.contents) == 0:
    self.delete()
```

**Comment by Griatch** [comment]
@mcarella  The `at_object_delete` method should definitely work, that is called at the beginning of the delete operation (just remember that `at_object_delete` must return `True` for deletion to continue, or it will not actually delete the object).

That said, it looks like your `dropanddie` solution should work too - you may want to put some `print` in there and check the server log to make sure your code is getting reached in the way you expect. Also look for tracebacks in the log, in case something goes wrong as part of the `move_to` operation the `self.delete()` line may not be reached.

**Comment by kovitikus** [comment, marked as answer]
You will commonly find `move_to` to fail if the associated hooks cannot be called properly. Another option is to attempt to silence these hooks.

Using `move_hooks=False` as a kwarg in your `move_to`, along with `quiet=True` results in the most quiet move possible.

**Reply by mcarella** [reply]
Supermega thanks! This worked :-)

**Reply by kovitikus** [reply]
@mcarella

Be sure to mark posts as the answer if it resolved your issue. This helps others know that the problem is solved and the answer may help them as well.
